### ISSUE TITLE
Inconsistent test case names

### ISSUE THREAD

**Issue by bradleyharden** [VUnit/vunit#531]
I have a test case that is parameterized by three test bench generics. When I add configurations to specify the generics, I use `str(generics)` as the configuration name (where `generics` is a dictionary).

Strangely, I find that the order of variables in the configuration name is different than the order of variables when listing tests using `./run.py -l`. As a consequence, if I copy and paste the test name from `./run.py -l`, VUnit will give a `No tests selected` warning. The only way to run such a test is to use a wildcard, or to print the configuration name from `run.py` and use it to create a name that filters correctly.

Furthermore, upon deeper investigation, I find that when running the following commands in `run.py`

```Python
modelsim = ui._create_simulator_if()
#ui._test_bench_list.create_tests(modelsim, ui._args.elaborate)
#ui._test_bench_list.create_tests(None, ui._args.elaborate)
print("\n".join(test_list.test_names))
```
I get different variable ordering in the test names depending on whether the simulator argument is `modelsim` or `None`. Moreover, neither of the two different names matches that of `./run.py -l`.

By the way, the order of variables in the file `test_name_to_path_mapping.txt` matches that of `./run.py -l`. I have been using that text file to display the waveforms from previous runs of Modelsim using `vsim -view`, but the inconsistency of test names also prevents me from doing so in this case.

I tried to dig through the code to find the source of the problem, but I was getting lost. @kraigher and @LarsAsplund, do you have any instincts about the cause of the problem?

**Comment by LukasVik** [comment]
Hey, so this one has been open and perhaps you've already solved it but here are a few pointers.

1. Python dictionaries are by design not ordered. Without seeing the exact code you're using, this could be the cause of what you're seeing. If you want it ordered you can sort them by key, or use OrderedDict.
2. Maybe you're aware but casting a dictionary to string does not yield a very nice string:
`>>> str(dict(a=10, b=5))
"{'a': 10, 'b': 5}"`
I would definitely not want quotation marks, braces, or spaces in my test case names.

**Comment by bradleyharden** [comment]
@LukasVik, yes I'm aware that they are not fundamentally ordered. However, I was under the impression that creating the exact same dictionary, with only strings as keys, would always yield the same order. It looks like this is true within a single execution, but it is not true across executions. String hash randomization was added in 2012 to prevent denial of service attacks based on hash collisions (https://mail.python.org/pipermail/python-announce-list/2012-March/009394.html). That was the source of my problem.

With regard to your second point, VUnit seems to handle special characters in test names just fine. Certainly spaces are fine, because almost every test case name in the VUnit code base features them. And in my experience, quotation marks, braces and colons have not been a problem either.

**Comment by 1138-4EB** [comment]
@bradleyharden, might VUnit be enhanced by ensuring the the order is deterministic (using `OrderedDict` as @LukasVik said)? Or can you get consistent naming now that you understand the source of the problem?

**Comment by bradleyharden** [comment]
The issue isn't really with VUnit. VUnit will use whatever string name is provided by the user. 

Because I saw the same name with different ordering, I assumed that VUnit was separately generating a name for the test internally somewhere. I tried to find where that was happening within the VUnit code base, but I couldn't. Now I know why. It wasn't happening.

As long as the user-generated test case name is consistent across invocations, everything VUnit does should be fine. Now that I know the source of the problem, I will generate the names differently.
